include/miral/miral/x11_support.h

```diff
@@ -32,6 +32,7 @@ class X11Support
     void operator()(mir::Server& server) const;
 
     X11Support();
+    explicit X11Support(bool x11_enabled);
```

Comment:
I wonder if a [named constructor](https://isocpp.org/wiki/faq/ctors#named-ctor-idiom) would be clearer? I don't see much point in `X11Support{false}`

Reply:
Indeed. So, maybe something like that?

```
static X11Support x11_disabled_by_default();
static X11Support x11_enabled_by_default();
```

Would it be better to keep the default constructor as public for backwards compatibility? The downside is that it won't be clear to the user that it defaults to x11 as disabled.

Reply:
Think of the client code:

```c++
    ...
    X11Support::x11_enabled_by_default(),
    ...
```

would be rather tautological.

And, yes. We don't need to break ABI compatibility by removing the current constructor. (That would require an .so name change)

Reply:
I agree that it seems redundant, but `X11Support` doesn't imply X11 is enabled. In the current context, having support for X11 simply means providing a command line option to the user. Perhaps `X11Support` should more accurately be named `X11EnableOption` which could then be `disabled_by_default` or `enabled_by_default` based on the chosen named constructor. However, as we plan to keep the default constructor and only add a named one for enabling X11, I understand that we will end up with something like this:

- `X11Support()`: the default constructor, which doesn't enable X11. The "support" is there (i.e. the command line option), but X11 is disabled unless the user states otherwise.
- `X11Support::enabled_by_default()`: named constructor, which adds the command line option and also enables X11 by default.

I am open to suggestions.

Reply:
Oh, I must correct myself! Upon revisiting the code, I noticed that `X11Support` includes additional configuration options (`x11-scale`, `x11-displayfd`, `xwayland-path`): using `X11EnableOption` to describe its use would be incorrect.

Given that `X11Support` encompasses a set of initialization options with default values, I wonder if a _named parameter idiom_ would enhance readability in this case. The client could write something like `X11Support().set_enabled_by_default()`, extensible to `X11Support().set_enabled_by_default().set_default_scale(2)...` if needed.

Another option would be to use a struct of default values and designated initializers; for instance, `X11Support({.enabled = true})`, extensible to `X11Support({.enabled = true, scale = 1.0f, ...})`.

What are your thoughts on these alternatives?

Reply:
I like the method chaining approach. Although I think dropping the "set_" wart, or replacing it with "with_" would be better:

```c++
    ...
    X11Support{}.default_to_enabled().default_to_scale(2),
    ...
```